lib/space_saving.c

```diff
@@ -0,0 +1,795 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+/*
+ * This Implementation is similar to the stream summary data structure decribed
+ * in the research paper on space saving algorithm by Ahmed Metwally,
+ * Divyakant Agrawal and Amr El Abbadi.
+ */
+
+/* 
+ * TODO #1 : Add a Tree based data structure such as AVL Tree or
+ * Van Emde Boas Tree to support sublinear bucket update operation.
+ */
+
+/* TODO #2 : Test the written code for correctness */
+
+#include <time.h>
+
+#include "space_saving.h"
+
+/* Set up the parameters of a counter bucket */
+static int
+setup_bucket_params(unsigned int socket_id, int bkt_size, int bkt_id,
+		int ip_ver, struct rte_hash_parameters *bkt_hash_params)
+{
+	int ret;
+
+	int key_len = ip_ver == ETHER_TYPE_IPv4 ?
+			sizeof(struct in_addr) : sizeof(struct in6_addr);
+
+	bkt_hash_params->name = NULL;
+	bkt_hash_params->entries = bkt_size;
+	bkt_hash_params->key_len = key_len;
+	bkt_hash_params->hash_func = DEFAULT_HASH_FUNC;
+	bkt_hash_params->hash_func_init_val = 0;
+	bkt_hash_params->reserved = 0;
+
+	char bkt_name[64];
+	ret = snprintf(bkt_name, sizeof(bkt_name), "Bucket_number_%d_%d",
+			ip_ver, bkt_id);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(bkt_name));
+
+	bkt_hash_params->name = bkt_name;
+	bkt_hash_params->socket_id = socket_id;
+
+	return ret;
+}
+
+/* Create a counter bucket */
+static int
+create_bucket(struct gatekeeper_if *iface, struct gk_config *gk_conf,
+		int bkt_id)
+{
+	int ret;
+	unsigned int socket_id;
+	struct rte_hash_parameters bkt_hash_params;
+	struct counter_bucket *ct_bkt =
+		(struct counter_bucket *)malloc(sizeof(struct counter_bucket));
+
+	socket_id = rte_lcore_to_socket_id(gk_conf->lcores[0]);
+
+	/*
+	 * TODO : How to determine max_size of a bucket ?
+	 * Currently it has been set to "10000" but final values has to be
+	 * updated after testing.
+	 */
+
+	if(ipv4_if_configured(iface)) {
+		ret = setup_bucket_params(socket_id, bkt_id, ETHER_TYPE_IPv4,
+						10000, &bkt_hash_params);
```

Comment:
Typically, please define a const value or make it configurable for a value needed to be tuned.

Reply:
Alignment issue: please make it one tab more than the above line.